### PR TITLE
add threadlocal api key to allow using public and internal APIs in a multithreaded server setting

### PR DESCRIPTION
**Pull request by dannygoldstein**
Description
-----------
The weave team is using `wandb` in a multithreaded server environment to make requests to the wandb api on behalf of authenticated W&B users. Because the server handles many requests concurrently we need a way of setting a single API key that all `InternalApi()` and `public.Api()` queries should use within the context of a single request without leaking state across requests. This is not possible with `Settings` or environment variable API key passing as these are scoped to processes not threads. Here we add a single thread-local state container `_ThreadLocalApiKey` which can be set by the weave server before processing each incoming request. All `wandb` API calls will use this key if no others are set from higher priority sources like the environment or settings. And because the storage is thread local there is no chance of bleeding the API key across requests. 

This works in the user process only and should not be relied on for calls in backend processes or threads like the sender thread. The motivation for adding this is that https://github.com/wandb/wandb/pull/3657 was not enough. Methods of the public API like `artifact()` actually instantiate `InternalApi()` and use that api to make calls to `wandb`. The explicit api_key from that PR is not passed to these intra-api instantiations, so we have to figure out a way to get the API key to the relevant APIs without a proliferation of explicit plumbing throughout the codebase. 

The new API key source we have added is the lowest priority except for the default. It is only used by the InternalApi and public.Api()

Testing
-------
How was this PR tested?

Tested on QA 

Checklist
-------
- Include reference to internal ticket "Fixes WB-NNNN" (and github issue "Fixes #NNNN" if applicable)

### CODE REVIEW

File: wandb/sdk/internal/internal_api.py [hunk 1]
```diff
@@ -53,6 +55,20 @@
 
 logger = logging.getLogger(__name__)
 
+
+# Context variable for setting API keys for internal and public apis thread-locally
+class _ThreadLocalApiKey(threading.local):
```
Comment (tssweeney) [inline wandb/sdk/internal/internal_api.py:60]: i wonder if we should call this `_ThreadLocalSettings` - where the only allowed setting right now is api_key?

**Review by tssweeney — APPROVED** [review]

File: wandb/sdk/internal/internal_api.py (hunk 1 above)
Comment (dannygoldstein) [reply wandb/sdk/internal/internal_api.py]: done

**Review by nickpenaranda — COMMENTED** [review]
If I am configuring a key for a specific thread I would expect that to override other less specific sources.  I understand why we chose to put it second-to-last, but to me this violates principle of least surprise

**Comment by dmitryduev** [comment]
@dannygoldstein the new testing infra for the sdk is in master, so you may want to start thinking about proper tests for this. (ping me if you need help with that)

**Comment by dmitryduev** [comment]
@dannygoldstein any updates on this?

**Pull request closed without merging**